grid2op/multi_agent/multiAgentEnv.py

```diff
@@ -0,0 +1,1070 @@
+        #     tmp_cls.mask_shunt
+        # ]
+        # tmp_cls.name_shunt = self._cent_env.name_shunt[
+        #     tmp_cls.mask_shunt
+        # ]
+        # tmp_cls.shunt_to_subid = self._cent_env.shunt_to_subid[
+        #     tmp_cls.mask_shunt
+        # ]
+
+        # # alarms #TODO
+        # tmp_cls.dim_alarms = 0
+        # tmp_cls.alarms_area_names = []
+        # tmp_cls.alarms_lines_area = {}
+        # tmp_cls.alarms_area_lines = []
+        
+        return tmp_cls
+        
+    
+    def _build_observation_spaces(self, global_obs = True):
+        """Build observation spaces from given domains for each agent
+
+        Args:
+            global_obs (bool, optional): The agent can observe the whole grid. Defaults to True.
+
+        Raises:
+            NotImplementedError: Local observations are not implemented yet
+        """
+        
+        if global_obs:
+            self.global_obs = True
+            for agent in self.agents:
+                self.observation_spaces[agent] = self._cent_env.observation_space
+        else:
+            raise NotImplementedError("Local observations are not available yet !")
+        #TODO Local observations
+        #for agent in self.agents: 
+        #    _cls_agent_action_space = ObservationSpace.init_grid(gridobj=self._subgrids_cls['observation'][agent], extra_name=agent)
+        #    self.observation_spaces[agent] = _cls_agent_action_space(
+        #        gridobj = self._subgrids_cls['observation'][agent],
+        #        env = self._cent_env,
+        #        rewardClass=self._cent_env._rewardClass,
+        #        observationClass=self._cent_env._observationClass,
+        #        actionClass=self._cent_env._actionClass,
+        #        #TODO following parameters
+        #        with_forecast=True,
+        #        kwargs_observation=None,
+        #    )
+    
+    def _build_action_spaces(self):
+        """Build action spaces from given domains for each agent
+        The action class is the same as 
+        """
+        for agent in self.agents: 
+            _cls_agent_action_space = ActionSpace.init_grid(gridobj=self._subgrids_cls['action'][agent], extra_name=agent)
+            self.action_spaces[agent] = _cls_agent_action_space(
+                gridobj=self._subgrids_cls['action'][agent],
+                actionClass=self._cent_env._actionClass,
+                legal_action=self._cent_env._game_rules.legal_action,
+            )
+    
+    
+    def _update_observations(self, observation : BaseObservation):
+        """Update agents' observations from the global observation given by the self._cent_env
+
+        Args:
+            observation (BaseObservation): _description_
+        """
+        if self.global_obs:
+            for agent in self.agents:
+                self.observations[agent] = observation.copy()
+        else:
+            #TODO
+            raise NotImplementedError("Local observations are not available yet !")
```

Comment:
Also, here, you suppose that you need the observation of the full grid to build the sub observation.

Do you really ?

You can do this with only the state of the backend ? Why do you chose to do that this way ? What does it entails ? What are the pros and cons ?

Reply:
For global observations, I think it's faster to give a copy of the observation computed and given by the _cent_env.step. Otherwise, we'll do the exact same computation n_agents times!